File: core/src/main/cpp/entityauth/RsaAuthenticationData.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <entityauth/EntityAuthenticationScheme.h>
#include <entityauth/RsaAuthenticationData.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <io/MslObject.h>
#include <MslEncodingException.h>
#include <MslError.h>
#include <util/MslUtils.h>

using namespace std;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace entityauth {

namespace {
/** Key entity identity. */
const string KEY_IDENTITY = "identity";
/** Key public key ID. */
const string KEY_PUBKEY_ID = "pubkeyid";
}

RsaAuthenticationData::RsaAuthenticationData(const string identity, const string pubkeyid)
    : EntityAuthenticationData(EntityAuthenticationScheme::RSA), identity_(identity), pubkeyid_(pubkeyid)
{
}

// note weird syntax to handle an exception thrown in initializer list
RsaAuthenticationData::RsaAuthenticationData(shared_ptr<MslObject> rsaAuthMo)
try
    : EntityAuthenticationData(EntityAuthenticationScheme::RSA)
    , identity_(rsaAuthMo->getString(KEY_IDENTITY))
	, pubkeyid_(rsaAuthMo->getString(KEY_PUBKEY_ID))
{
}
catch (const MslEncoderException& e)
{
    stringstream ss;
    ss << "rsa authdata " << rsaAuthMo;
    throw MslEncodingException(MslError::MSL_PARSE_ERROR, ss.str(), e);
}

shared_ptr<MslObject> RsaAuthenticationData::getAuthData(shared_ptr<MslEncoderFactory> factory,
        const MslEncoderFormat& ) const
{
    shared_ptr<MslObject> mo = factory->createObject();
    mo->put<string>(KEY_IDENTITY, identity_);
    mo->put<string>(KEY_PUBKEY_ID, pubkeyid_);
    return mo;
}

bool RsaAuthenticationData::equals(shared_ptr<const EntityAuthenticationData> obj) const
{
    if (this == obj.get()) return true;
    if (!instanceof<const RsaAuthenticationData>(obj.get())) return false;
    shared_ptr<const RsaAuthenticationData> that = dynamic_pointer_cast<const RsaAuthenticationData>(obj);
    return EntityAuthenticationData::equals(obj) && identity_ == that->identity_ && pubkeyid_ == that->pubkeyid_;
}

bool operator==(const RsaAuthenticationData& a, const RsaAuthenticationData& b)
{
	shared_ptr<const RsaAuthenticationData> ap(&a, &MslUtils::nullDeleter<RsaAuthenticationData>);
	shared_ptr<const RsaAuthenticationData> bp(&b, &MslUtils::nullDeleter<RsaAuthenticationData>);
	return ap->equals(bp);
}

}}} // namespace netflix::msl::entityauth

File: core/src/main/javascript/entityauth/RsaAuthenticationData.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>RSA asymmetric keys entity authentication data.</p>
 *
 * <p>
 * {@code {
 *   "#mandatory" : [ "identity", "pubkeyid" ],
 *   "identity" : "string",
 *   "pubkeyid" : "string"
 * }} where:
 * <ul>
 * <li>{@code identity} is the entity identity</li>
 * <li>{@code pubkeyid} is the identity of the RSA public key associated with this identity</li>
 * </ul></p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
    "use strict";

    var EntityAuthenticationData = require('../entityauth/EntityAuthenticationData.js');
    var EntityAuthenticationScheme = require('../entityauth/EntityAuthenticationScheme.js');
    var AsyncExecutor = require('../util/AsyncExecutor.js');
    var MslEncoderException = require('../io/MslEncoderException.js');
    var MslEncodingException = require('../MslEncodingException.js');
    var MslError = require('../MslError.js');

    /**
     * Key entity identity.
     * @const
     * @type {string}
     */
    var KEY_IDENTITY = "identity";
    /**
     * Key public key ID.
     * @const
     * @type {string}
     */
    var KEY_PUBKEY_ID = "pubkeyid";

    var RsaAuthenticationData = module.exports = EntityAuthenticationData.extend({
        /**
         * <p>Construct a new public key authentication data instance from the
         * specified entity identity and public key ID.</p>
         *
         * @param {string} identity the entity identity.
         * @param {string} pubkeyid the public key ID.
         */
        init: function init(identity, pubkeyid) {
            init.base.call(this, EntityAuthenticationScheme.RSA);

            // The properties.
            var props = {
                identity: { value: identity, writable: false, configurable: false },
                publicKeyId: { value: pubkeyid, writable: false, configurable: false },
            };
            Object.defineProperties(this, props);
        },

        /** @inheritDoc */
        getIdentity: function getIdentity() {
            return this.identity;
        },

        /** @inheritDoc */
        getAuthData: function getAuthData(encoder, format, callback) {
            AsyncExecutor(callback, function() {
                var mo = encoder.createObject();
                mo.put(KEY_IDENTITY, this.identity);
                mo.put(KEY_PUBKEY_ID, this.publicKeyId);
                return mo;
            }, this);
        },

        /** @inheritDoc */
        equals: function equals(that) {
            if (this === that) return true;
            if (!(that instanceof RsaAuthenticationData)) return false;
            return (equals.base.call(this, that) && this.identity == that.identity && this.publicKeyId == that.publicKeyId);
        },
    });

    /**
     * Construct a new RSA asymmetric keys authentication data instance from the
     * provided MSL object.
     *
     * @param {MslObject} rsaAuthMo the authentication data MSL object.
     * @return the authentication data.
     * @throws MslEncodingException if there is an error parsing the entity
     *         authentication data.
     */
    var RsaAuthenticationData$parse = function RsaAuthenticationData$parse(rsaAuthMo) {
        try {
            var identity = rsaAuthMo.getString(KEY_IDENTITY);
            var pubkeyid = rsaAuthMo.getString(KEY_PUBKEY_ID);
            return new RsaAuthenticationData(identity, pubkeyid);
        } catch (e) {
            if (e instanceof MslEncoderException)
                throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "RSA authdata" + rsaAuthMo);
            throw e;
        }
    };
    
    // Exports.
    module.exports.parse = RsaAuthenticationData$parse;
})(require, (typeof module !== 'undefined') ? module : mkmodule('RsaAuthenticationData'));
